### PR TITLE
Webhook for PD upgrader

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/issues/992

### What is changed and how does it work?
Tidb-Operator will only need to edit statefulset spec during pd upgrading and to edit replicas during pd scale-in. The webhook server will ensure that each pd pod will be safely deleted.

And I also add one more component as an init job to auto-create CA cert and csr for webhook configuration which would be much user-friendly.

Currently I package `tidb-operator-initializer` in the same docker image as others and install `openssl` and `kubectl` in `tidb-operator` image which `initializer` use these in the script to manage CA cert. 

Packaging all necessary process into 2 images would need us to change some e2e test script which would have huge influence for other e2e test. 

After we moved our e2e CI script into code repo instead of jenkins server, I would pakcage `tidb-operator-initializer` in a different image so that other e2e test won't be influenced.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has Helm charts change
 - Has Go code change

Side effects
N/A

Related changes

 ```release-note
Kubernetes 1.9.0 or above with the admissionregistration.k8s.io/v1beta1 API enabled. 
 ```

### CODE REVIEW

**Comment by Yisaer** [comment]
This is a draft for webhook refatcto design for now, I will add test case after discussion.

File: charts/tidb-operator/values.yaml [hunk 1]
```diff
@@ -124,3 +124,13 @@ apiserver:
...
+admissionControllerWebhook:
+  resources:
+    limits:
+      cpu: 250m
+      memory: 150Mi
+    requests:
+      cpu: 80m
+      memory: 50Mi
+  ## Interval Hour to Refresh CA Cert
+  certRefreshIntervalHour: 336
```
Comment (Yisaer) [inline charts/tidb-operator/values.yaml]: AdmissionController will refresh CA cert before `certRefreshIntervalHour` hour on its expireDate.

File: cmd/admission-controller/main.go [hunk 2]
```diff
@@ -68,7 +126,66 @@ func main() {
...
+		}
+	}
+	glog.Infof("cache of informer factories sync successfully")
+
+	if err := webhookServer.Run(); err != nil {
+		glog.Errorf("stop http server %v", err)
+	}
+
+	/*
+	* currently we only use one replica for admission-controller-webhook
```
Comment (Yisaer) [inline cmd/admission-controller/main.go]: Currently we just set replica as 1 for admissioncontroller so we don't have lease lock from apiServer as we are using Lister here.

File: pkg/util/certUtil.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+func Checksum(src []byte) string {
+	h := sha256.New()
+	h.Write(src)
+	out := h.Sum(nil)
+	return hex.EncodeToString(out)
+}
+
+func GenerateRSAKeyPair(namespace, serviceName string) ([]byte, []byte, error) {
```
Comment (Yisaer) [inline pkg/util/certUtil.go]: CertUtil help us generate CA Cert for a certain service with ten years default validity period

File: pkg/webhook/pod/pods.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+	return &PodAdmissionControl{
+		kubeCli:     kubeCli,
+		operatorCli: operatorCli,
+		pvcControl:  PVCControl,
+		pdControl:   PdControl,
+	}
+}
+
+func (pc *PodAdmissionControl) AdmitPods(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
```
Comment (Yisaer) [inline pkg/webhook/pod/pods.go]: AdmitPods will only check AdmissionReview with Delete Operation.

File: pkg/webhook/pod/pods.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+	// If there was only one replica for this statefulset,admit to delete it.
+	if *ownerStatefulSet.Spec.Replicas == 1 && ordinal == 0 {
+		glog.Infof("tc[%s/%s]'s pd only have one pod[%s/%s],admit to delete it.", namespace, tcName, namespace, name)
+		return util.ARSuccess()
+	}
+
+	if l.IsPD() {
+		pdClient := pc.pdControl.GetPDClient(pdapi.Namespace(namespace), tcName, tc.Spec.EnableTLSCluster)
+		return pc.admitDeletePdPods(pod, ownerStatefulSet, tc, pdClient)
```
Comment (Yisaer) [inline pkg/webhook/pod/pods.go]: Currently we only check Delete AdmissionReview For PD Pod.

File: tests/cmd/e2e/main.go [hunk 6]
```diff
@@ -133,7 +128,6 @@ func main() {
 		oa.CheckTidbClusterStatusOrDie(cluster)
 
 		// upgrade
-		oa.RegisterWebHookAndServiceOrDie(certCtx, ocfg)
```
Comment (Yisaer) [inline tests/cmd/e2e/main.go]: Remove e2e webhook.

File: pkg/initializer/initializer.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+// update Webhook server to make sure the latest secret would be used.
+func (initializer *Initializer) updateWebhookServer(namespace string, secret *core.Secret) error {
+
+	server, err := initializer.kubeCli.ExtensionsV1beta1().Deployments(namespace).Get(admissionWebhookServerName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	server.Spec.Template.Annotations = map[string]string{
+		"checksum/config": Checksum(secret.Data["cert.pem"]),
```
Comment (Yisaer) [inline pkg/initializer/initializer.go]: We update deployment by updating its annotations to get the latest cert secret.

**Comment by Yisaer** [comment]
@tennix @weekface @cofyc @aylei  PTAL.

**Comment by cofyc** [comment]
Can we divide the whole process into several steps?

#### implement the webhook controller and can use the webhook in tidb-operator optionally 

1) we can use hardcoded self-signed certs in e2e tests
     - it seems that we can use a custom ca bundle in `clientConfig`: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#service-reference
2) add a switch to disable some logic in controller-manager to work with webhook admission controller

in this step, we focus on webhook logic

#### deploy/configure webhook controller

request/sign/configure/renew cert for webhook controller

in this step, we focus on webhook deployment and management

#### use the webhook controller by default

when it's stable, we can switch to use webhook controller by default

if it's production-ready, we can remove the old logic in the controller and require the webhook to exist in our deployment

File: pkg/webhook/pod/pd_deleter.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+				glog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
+				return util.ARFail(err)
+			}
+		}
+		glog.Infof("pd pod[%s/%s] is not member of tc[%s/%s],admit to delete", namespace, name, namespace, tcName)
+		return util.ARSuccess()
+	}
+
+	if isOutOfOrdinal {
+		err = pdClient.DeleteMember(name)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/webhook/pod/pd_deleter.go]: `DeleteMember` will auto transfer pd leader if it was.

**Comment by tennix** [comment]
Is this PR still needed after #1101 is merged?

**Comment by Yisaer** [comment]
This request is no more needed now.

**Pull request closed without merging**